ee/backend/app/src/main/kotlin/io/tolgee/ee/service/prompt/PromptServiceEeImpl.kt

```diff
@@ -0,0 +1,461 @@
+          val id = match.groups[2]!!.value.toLong() // number
+          val screenshot = key?.keyScreenshotReferences?.find { it.screenshot.id == id }?.screenshot
+          if (screenshot == null) {
+            null
+          } else {
+            val file =
+              if (size == "full") {
+                screenshot.filename
+              } else {
+                screenshot.middleSizedFilename ?: screenshot.filename
+              }
+
+            val filePath = screenshotService.getScreenshotPath(file)
+
+            lateinit var image: ByteArray
+
+            if (screenshot.keyScreenshotReferences.find { it.key.id == key?.id } !== null) {
+              val converter =
+                ImageConverter(
+                  ByteArrayInputStream(
+                    fileStorage.readFile(filePath),
+                  ),
+                )
+              image = converter.highlightKeys(screenshot, listOf(key.id)).toByteArray()
+            } else {
+              image = fileStorage.readFile(filePath)
+            }
+
+            LlmParams.Companion.LlmMessage(
+              type = LlmParams.Companion.LlmMessageType.IMAGE,
+              image = Base64.getEncoder().encodeToString(image),
+            )
+          }
+        } else {
+          LlmParams.Companion.LlmMessage(
+            type = LlmParams.Companion.LlmMessageType.TEXT,
+            text = it,
+          )
+        }
+      }
+    return LlmParams(messages, shouldOutputJson, priority)
+  }
+
+  // Helper function to split and keep matches
+  fun Regex.splitWithMatches(input: String): List<String> {
+    val result = mutableListOf<String>()
+    var lastIndex = 0
+
+    this.findAll(input).forEach { match ->
+      // Add text before the match if exists
+      if (match.range.first > lastIndex) {
+        result.add(input.substring(lastIndex, match.range.first))
+      }
+      // Add the match itself
+      result.add(match.value)
+      lastIndex = match.range.last + 1
+    }
+
+    // Add remaining text after last match
+    if (lastIndex < input.length) {
+      result.add(input.substring(lastIndex))
+    }
+
+    if (result.size == 0) {
+      result.add("")
+    }
+
+    return result
+  }
+
+  fun runPromptWithoutChargingCredits(
+    organizationId: Long,
+    params: LlmParams,
+    provider: String,
+    attempts: List<Int>? = null,
+  ): PromptResult {
+    val result =
+      try {
+        providerService.callProvider(organizationId, provider, params, attempts)
+      } catch (e: RestClientException) {
+        throw BadRequestException(Message.LLM_PROVIDER_ERROR, listOf(e.message), e)
+      } catch (e: TranslationApiRateLimitException) {
+        throw BadRequestException(Message.LLM_RATE_LIMITED, listOf(e.message), e)
+      }
+
+    result.parsedJson =
+      try {
+        jacksonObjectMapper().readValue<JsonNode>(result.response)
+      } catch (e: JsonProcessingException) {
+        null
+      }
+    return result
+  }
+
+  fun runPrompt(
```

Comment:
Consider moving the prompt running metod and the code related to it to separate component. It is long and might make sense to sprit to provider entity storing logic and the propmt running logic.

Reply:
It's a part of public API, I don't think it make sense